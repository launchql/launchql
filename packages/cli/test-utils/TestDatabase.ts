import { PgConfig } from 'pg-env';

export interface TestDatabase {
  name: string;
  config: PgConfig;
  
  query(sql: string, params?: any[]): Promise<any>;
  exists(type: 'schema' | 'table', name: string): Promise<boolean>;
  getDeployedChanges(): Promise<any[]>;
  getMigrationState(): Promise<{
    changes: any[];
    events: any[];
    changeCount: number;
    eventCount: number;
  }>;
  getDependencies(project: string, changeName: string): Promise<string[]>;
  close(): Promise<void>;
}
