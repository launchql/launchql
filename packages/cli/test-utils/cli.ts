import readline from 'readline';
import { Readable, Transform, Writable } from 'stream';
import stripAnsi from 'strip-ansi';

import { CLIOptions } from 'inquirerer';

export const KEY_SEQUENCES = {
    ENTER: '\u000d',
    UP_ARROW: '\u001b[A',
    DOWN_ARROW: '\u001b[B',
    SPACE: ' '
};

function humanizeKeySequences(data: string): string {
    const keyMap: { [key: string]: string } = {
        '\u000d': '<ENTER>',
        '\u001b[A': '<UP_ARROW>',
        '\u001b[B': '<DOWN_ARROW>',
        ' ': '<SPACE>'
    };

    return data.replace(/[\u000d\u001b[A\u001b[B ]/g, (match) => keyMap[match] ?? match);
}
interface InputResponse {
    type: 'key' | 'read';
    value: string;
}

interface MockReadline {
    question: (questionText: string, cb: (input: string) => void) => void;
    close: () => void;
}

export interface TestEnvironment {
    options: Partial<CLIOptions>;
    mockInput: Readable;
    mockOutput: Writable;
    writeResults: string[];
    transformResults: string[];
    enqueueInputResponse: (input: InputResponse) => void;
}

function setupReadlineMock(inputQueue: InputResponse[], currentInputIndex: number): void {
    readline.createInterface = jest.fn().mockReturnValue({
        question: (questionText: string, cb: (input: string) => void) => {
            const nextInput = inputQueue[currentInputIndex++];
            if (nextInput && nextInput.type === 'read') {
                setTimeout(() => cb(nextInput.value), 1); // simulate readline 1ms
            }
        },
        close: jest.fn(),
    } as MockReadline);
}

export function setupTests(): () => TestEnvironment {
    let options: Partial<CLIOptions>;
    let mockWrite: jest.Mock;
    let mockInput: Readable;
    let mockOutput: Writable;
    let transformStream: Transform;

    let writeResults: string[] = [];
    let transformResults: string[] = [];

    let inputQueue: InputResponse[] = [];
    let currentInputIndex = 0;
    let lastScheduledTime = 0;


    const beforeEachSetup = (): TestEnvironment => {
        jest.clearAllMocks();
        mockWrite = jest.fn();
        writeResults = [];
        transformResults = [];

        mockInput = new Readable({ read(size) { } });
        (mockInput as any).setRawMode = jest.fn();

        mockOutput = new Writable({
            write: (chunk, encoding, callback) => {
                const str = chunk.toString();
                const humanizedStr = humanizeKeySequences(str);
                const cleanStr = stripAnsi(humanizedStr);
                writeResults.push(cleanStr);
                mockWrite(str);
                callback();
            }
        });

        // mock I/O streams so we can keep TTY for testing and CI/CD ðŸŽ¨
        options = {
            noTty: false,
            input: mockInput,
            output: mockOutput,
            minimistOpts: {
                alias: {
                    v: 'version'
                }
            }
        };

        transformStream = new Transform({
            transform(chunk, encoding, callback) {
                const data = chunk.toString();
                const humanizedData = humanizeKeySequences(data);
                const cleanData = stripAnsi(humanizedData);
                transformResults.push(cleanData);
                this.push(chunk);
                callback();
            }
        });

        setupReadlineMock(inputQueue, currentInputIndex);
        mockInput.pipe(transformStream);

        const enqueueInputResponse = (input: InputResponse) => {
            lastScheduledTime += 1;

            if (input.type === 'key') {
                setTimeout(() => mockInput.push(input.value), lastScheduledTime);
            } else {
                inputQueue.push(input);  // We assume that handling read inputs remains the same unless you need to delay these as well
            }
        };


        return {
            options,
            mockInput,
            mockOutput,
            writeResults,
            transformResults,
            enqueueInputResponse
        };
    };

    return beforeEachSetup;
}
