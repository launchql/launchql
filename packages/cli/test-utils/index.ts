export * from './cli';
export * from './fixtures';
export * from './CLIDeployTestFixture';
export * from './TestDatabase';
