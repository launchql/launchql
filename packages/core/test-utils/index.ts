export { TestPlan } from './TestPlan';
export { CoreDeployTestFixture } from './CoreDeployTestFixture';
export { TestFixture } from './TestFixture';

export { MigrateTestFixture } from './MigrateTestFixture';
export { TestDatabase } from './TestDatabase';
export { MigrateTestChange } from './MigrateTestChange';

export {
  FIXTURES_PATH,
  getFixturePath,
  cleanText,
  teardownAllPools,
  closeDatabasePools
} from './utils';
