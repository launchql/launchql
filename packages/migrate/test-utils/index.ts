import { MigrateConfig } from '../src/types';
import { LaunchQLMigrate } from '../src/client';
import { mkdtempSync, rmSync, cpSync, writeFileSync, mkdirSync } from 'fs';
import { join, resolve } from 'path';
import { tmpdir } from 'os';
import { getPgPool, teardownPgPools } from 'pg-cache';
import { getPgEnvOptions, PgConfig } from 'pg-env';
import { Pool } from 'pg';

export const FIXTURES_PATH = resolve(__dirname, '../../../__fixtures__/migrate');

// Global teardown function to be called after all tests
export async function teardownAllPools(): Promise<void> {
  await teardownPgPools();
}

// Helper to close specific database pools
export async function closeDatabasePools(databases: string[]): Promise<void> {
  // For now, we'll rely on the global teardown
  // Individual database pool cleanup can cause issues with pg-cache
}

export interface TestDatabase {
  name: string;
  config: MigrateConfig;
  query(sql: string, params?: any[]): Promise<any>;
  exists(type: 'schema' | 'table', name: string): Promise<boolean>;
  getDeployedChanges(): Promise<any[]>;
  getDependencies(project: string, changeName: string): Promise<string[]>;
  close(): Promise<void>;
}

export interface Change {
  name: string;
  dependencies?: string[];
  timestamp?: string;
  planner?: string;
  email?: string;
  comment?: string;
}

export class MigrateTestFixture {
  private tempDirs: string[] = [];
  private databases: TestDatabase[] = [];
  private dbCounter = 0;
  private pools: Pool[] = [];

  async setupTestDatabase(): Promise<TestDatabase> {
    const dbName = `test_migrate_${Date.now()}_${Math.random().toString(36).substring(2, 8)}_${this.dbCounter++}`;
    
    // Get base config from environment using pg-env
    const baseConfig = getPgEnvOptions({
      database: 'postgres'
    });
    
    // Create database using admin pool
    const adminPool = getPgPool(baseConfig);
    await adminPool.query(`CREATE DATABASE "${dbName}"`);
    
    // Get config for the new test database
    const pgConfig = getPgEnvOptions({
      database: dbName
    });
    
    const config: MigrateConfig = {
      host: pgConfig.host,
      port: pgConfig.port,
      user: pgConfig.user,
      password: pgConfig.password,
      database: pgConfig.database
    };

    // Initialize migrate schema
    const migrate = new LaunchQLMigrate(config);
    await migrate.initialize();

    // Get pool for test database operations
    const pool = getPgPool(pgConfig);
    this.pools.push(pool);

    const db: TestDatabase = {
      name: dbName,
      config,
      
      async query(sql: string, params?: any[]) {
        return pool.query(sql, params);
      },

      async exists(type: 'schema' | 'table', name: string) {
        if (type === 'schema') {
          const result = await pool.query(
            `SELECT EXISTS (
              SELECT 1 FROM information_schema.schemata 
              WHERE schema_name = $1
            ) as exists`,
            [name]
          );
          return result.rows[0].exists;
        } else {
          const [schema, table] = name.includes('.') ? name.split('.') : ['public', name];
          const result = await pool.query(
            `SELECT EXISTS (
              SELECT 1 FROM information_schema.tables 
              WHERE table_schema = $1 AND table_name = $2
            ) as exists`,
            [schema, table]
          );
          return result.rows[0].exists;
        }
      },

      async getDeployedChanges() {
        const result = await pool.query(
          `SELECT project, change_name, deployed_at 
           FROM launchql_migrate.changes 
           ORDER BY deployed_at`
        );
        return result.rows;
      },

      async getDependencies(project: string, changeName: string) {
        const result = await pool.query(
          `SELECT d.requires 
           FROM launchql_migrate.dependencies d
           JOIN launchql_migrate.changes c ON c.change_id = d.change_id
           WHERE c.project = $1 AND c.change_name = $2`,
          [project, changeName]
        );
        return result.rows.map((row: any) => row.requires);
      },

      async close() {
        // Don't close the pool here as it's managed by pg-cache
        // Just mark this database as closed
      }
    };

    this.databases.push(db);
    return db;
  }

  setupFixture(fixtureName: string): string {
    const originalPath = join(FIXTURES_PATH, fixtureName);
    const tempDir = mkdtempSync(join(tmpdir(), 'migrate-test-'));
    const fixturePath = join(tempDir, fixtureName);
    
    cpSync(originalPath, fixturePath, { recursive: true });
    this.tempDirs.push(tempDir);
    
    return fixturePath;
  }

  createPlanFile(project: string, changes: Change[]): string {
    const tempDir = mkdtempSync(join(tmpdir(), 'migrate-test-'));
    this.tempDirs.push(tempDir);

    const lines = [
      '%syntax-version=1.0.0',
      `%project=${project}`,
      `%uri=https://github.com/test/${project}`,
      ''
    ];

    for (const change of changes) {
      let line = change.name;
      
      if (change.dependencies && change.dependencies.length > 0) {
        line += ` [${change.dependencies.join(' ')}]`;
      }
      
      line += ` ${change.timestamp || new Date().toISOString().replace(/\.\d{3}Z$/, 'Z')}`;
      line += ` ${change.planner || 'test'}`;
      line += ` <${change.email || 'test@example.com'}>`;
      
      if (change.comment) {
        line += ` # ${change.comment}`;
      }
      
      lines.push(line);
    }

    const planPath = join(tempDir, 'sqitch.plan');
    writeFileSync(planPath, lines.join('\n'));
    
    return tempDir;
  }

  createScript(dir: string, type: 'deploy' | 'revert' | 'verify', name: string, content: string): void {
    const scriptDir = join(dir, type);
    mkdirSync(scriptDir, { recursive: true });
    writeFileSync(join(scriptDir, `${name}.sql`), content);
  }

  async cleanup(): Promise<void> {
    // Close all test database connections first
    const dbNames = this.databases.map(db => db.name);
    
    // Close all pools for test databases
    await closeDatabasePools(dbNames);
    
    // Get admin pool for database cleanup
    // const adminConfig = getPgEnvOptions({
    //   database: 'postgres'
    // });
    // const adminPool = getPgPool(adminConfig);

    // Small delay to ensure connections are closed
    await new Promise(resolve => setTimeout(resolve, 50));

    for (const db of this.databases) {
      try {
        // Drop the database
        // await adminPool.query(`DROP DATABASE IF EXISTS "${db.name}"`);
      } catch (e) {
        // Ignore errors - database might have active connections
      }
    }

    // Remove temporary directories
    for (const dir of this.tempDirs) {
      try {
        rmSync(dir, { recursive: true, force: true });
      } catch (e) {
        // Ignore errors during cleanup
      }
    }
  }
}

// Helper class for building plan files programmatically
export class PlanBuilder {
  private project: string;
  private changes: Change[] = [];

  constructor(project: string) {
    this.project = project;
  }

  addChange(name: string, options: Partial<Change> = {}): this {
    this.changes.push({
      name,
      ...options
    });
    return this;
  }

  build(): { project: string; changes: Change[] } {
    return {
      project: this.project,
      changes: this.changes
    };
  }

  toString(): string {
    const lines = [
      '%syntax-version=1.0.0',
      `%project=${this.project}`,
      `%uri=https://github.com/test/${this.project}`,
      ''
    ];

    for (const change of this.changes) {
      let line = change.name;
      
      if (change.dependencies && change.dependencies.length > 0) {
        line += ` [${change.dependencies.join(' ')}]`;
      }
      
      line += ` ${change.timestamp || new Date().toISOString().replace(/\.\d{3}Z$/, 'Z')}`;
      line += ` ${change.planner || 'test'}`;
      line += ` <${change.email || 'test@example.com'}>`;
      
      if (change.comment) {
        line += ` # ${change.comment}`;
      }
      
      lines.push(line);
    }

    return lines.join('\n');
  }
}