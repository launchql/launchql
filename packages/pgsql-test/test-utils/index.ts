import path from 'path';
import fs from 'fs';
