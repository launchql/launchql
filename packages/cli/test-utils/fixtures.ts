import fs from 'fs';
import os from 'os';
import path from 'path';
import { sync as glob } from 'glob';
import { Inquirerer } from 'inquirerer';
import { commands } from '../src/commands';
import { ParsedArgs } from 'minimist';
import { TestEnvironment, setupTests } from './cli';

const { mkdtempSync, rmSync, cpSync } = fs;

export const FIXTURES_PATH = path.resolve(__dirname, '../../../__fixtures__');

export const getFixturePath = (...paths: string[]) =>
  path.join(FIXTURES_PATH, ...paths);

export class TestFixture {
  readonly tempDir: string;
  readonly tempFixtureDir: string;
  readonly getFixturePath: (...paths: string[]) => string;
  private environment: TestEnvironment;

  constructor(...fixturePath: string[]) {
    this.tempDir = mkdtempSync(path.join(os.tmpdir(), 'launchql-test-'));

    if (fixturePath.length > 0) {
      const originalFixtureDir = getFixturePath(...fixturePath);
      this.tempFixtureDir = path.join(this.tempDir, ...fixturePath);
      cpSync(originalFixtureDir, this.tempFixtureDir, { recursive: true });
    } else {
      this.tempFixtureDir = this.tempDir;
    }

    this.getFixturePath = (...paths: string[]) =>
      path.join(this.tempFixtureDir, ...paths);

    this.environment = setupTests()();
  }

  fixturePath(...paths: string[]) {
    return path.join(this.tempFixtureDir, ...paths);
  }

  cleanup() {
    rmSync(this.tempDir, { recursive: true, force: true });
  }

  parseScript(script: string): Array<{ command: string; args: ParsedArgs }> {
    return script
      .trim()
      .split('\n')
      .filter(line => line.trim() && !line.trim().startsWith('#'))
      .map(line => {
        const parts = line.trim().split(/\s+/);
        const [, command, ...argParts] = parts; // Skip 'lql'
        
        const args: ParsedArgs = {
          _: [command],
          cwd: this.tempFixtureDir,
          recursive: true,
          yes: true, // Auto-confirm prompts
          tx: true,  // Use transactions
          fast: false // Don't use fast deployment
        };

        for (let i = 0; i < argParts.length; i++) {
          const arg = argParts[i];
          if (arg.startsWith('--')) {
            const key = arg.slice(2);
            
            if (key.startsWith('no-')) {
              const booleanKey = key.slice(3); // Remove 'no-' prefix
              args[booleanKey] = false;
            } else if (i + 1 < argParts.length && !argParts[i + 1].startsWith('--')) {
              args[key] = argParts[++i];
            } else {
              args[key] = true;
            }
          }
        }

        return { command, args };
      });
  }

  async runCmd(argv: ParsedArgs) {
    const {
      mockInput,
      mockOutput,
      writeResults,
      transformResults
    } = this.environment;

    const prompter = new Inquirerer({
      input: mockInput,
      output: mockOutput,
      noTty: true
    });

    // @ts-ignore
    const result = await commands(argv, prompter, {});

    return {
      result,
      argv,
      writeResults,
      transformResults
    };
  }

  async getSnapshottableResults(commands: Array<{ command: string; args: ParsedArgs }>) {
    const results: Array<{ command: string; args: ParsedArgs & { cwd: string }; success: boolean }> = [];
    for (const { command, args } of commands) {
      const result = await this.runCmd(args);
      results.push({
        command,
        args: { ...args, cwd: '<CWD>' }, // Normalize cwd for snapshot
        success: result.result !== null && result.result !== undefined
      });
    }
    return results;
  }
}
