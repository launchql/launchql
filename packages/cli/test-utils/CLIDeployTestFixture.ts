import { LaunchQLMigrate } from '@launchql/core';
import { Pool } from 'pg';
import { getPgPool } from 'pg-cache';
import { getPgEnvOptions } from 'pg-env';
import { ParsedArgs } from 'minimist';
import { Inquirerer, CLIOptions } from 'inquirerer';

import { TestFixture } from './fixtures';
import { TestDatabase } from './TestDatabase';
import { commands } from '../src/commands';

export class CLIDeployTestFixture extends TestFixture {
  private databases: TestDatabase[] = [];
  private dbCounter = 0;
  private pools: Pool[] = [];

  constructor(...fixturePath: string[]) {
    super(...fixturePath);
  }


  async setupTestDatabase(): Promise<TestDatabase> {
    const dbName = `test_cli_${Date.now()}_${Math.random().toString(36).substring(2, 8)}_${this.dbCounter++}`;
    
    // Get base config from environment using pg-env
    const baseConfig = getPgEnvOptions({
      database: 'postgres'
    });
    
    // Create database using admin pool
    const adminPool = getPgPool(baseConfig);
    try {
      await adminPool.query(`CREATE DATABASE "${dbName}"`);
    } catch (e) {
      if (e && typeof e === 'object' && 'errors' in e && Array.isArray((e as any).errors)) {
        for (const err of (e as any).errors) {
          console.error('AggregateError item:', err);
        }
      } else {
        console.error('Test failure:', e);
      }
      throw e;
    }
    
    // Get config for the new test database
    const pgConfig = getPgEnvOptions({
      database: dbName
    });
    
    const config = {
      host: pgConfig.host,
      port: pgConfig.port,
      user: pgConfig.user,
      password: pgConfig.password,
      database: pgConfig.database
    };

    // Initialize migrate schema
    const migrate = new LaunchQLMigrate(config);
    await migrate.initialize();

    // Get pool for test database operations
    const pool = getPgPool(pgConfig);
    this.pools.push(pool);

    const db: TestDatabase = {
      name: dbName,
      config,
      
      async query(sql: string, params?: any[]) {
        return pool.query(sql, params);
      },

      async exists(type: 'schema' | 'table', name: string) {
        if (type === 'schema') {
          const result = await pool.query(
            `SELECT EXISTS (
              SELECT 1 FROM information_schema.schemata 
              WHERE schema_name = $1
            ) as exists`,
            [name]
          );
          return result.rows[0].exists;
        } else {
          const [schema, table] = name.includes('.') ? name.split('.') : ['public', name];
          const result = await pool.query(
            `SELECT EXISTS (
              SELECT 1 FROM information_schema.tables 
              WHERE table_schema = $1 AND table_name = $2
            ) as exists`,
            [schema, table]
          );
          return result.rows[0].exists;
        }
      },

      async getDeployedChanges() {
        const result = await pool.query(
          `SELECT package, change_name, deployed_at 
           FROM launchql_migrate.changes 
           ORDER BY deployed_at`
        );
        return result.rows;
      },

      async getMigrationState() {
        const changes = await pool.query(`
          SELECT package, change_name, script_hash, deployed_at
          FROM launchql_migrate.changes 
          ORDER BY deployed_at
        `);
        
        const events = await pool.query(`
          SELECT package, change_name, event_type, occurred_at, error_message, error_code
          FROM launchql_migrate.events 
          ORDER BY occurred_at
        `);
        
        const sanitizedEvents = events.rows;
        
        // Remove timestamps from objects for consistent snapshots
        const cleanChanges = changes.rows.map(({ deployed_at, ...change }) => change);
        const cleanEvents = sanitizedEvents.map(({ occurred_at, ...event }) => event);
        
        return {
          changes: cleanChanges,
          events: cleanEvents,
          changeCount: cleanChanges.length,
          eventCount: cleanEvents.length
        };
      },

      async getDependencies(packageName: string, changeName: string) {
        const result = await pool.query(
          `SELECT d.requires 
           FROM launchql_migrate.dependencies d
           JOIN launchql_migrate.changes c ON c.change_id = d.change_id
           WHERE c.package = $1 AND c.change_name = $2`,
          [packageName, changeName]
        );
        return result.rows.map((row: any) => row.requires);
      },

      async close() {
        // Don't close the pool here as it's managed by pg-cache
        // Just mark this database as closed
      }
    };

    this.databases.push(db);
    return db;
  }


  async exec(commandString: string, variables: Record<string, string> = {}): Promise<any[]> {
    return this.runTerminalCommands(commandString, variables, true);
  }

  async runTerminalCommands(commandString: string, variables: Record<string, string> = {}, executeCommands: boolean = true): Promise<any[]> {
    const results: any[] = [];
    let currentDir = this.tempFixtureDir;
    
    const commands = commandString
      .split(/\n|;/)
      .map(cmd => cmd.trim())
      .filter(cmd => cmd.length > 0);
    
    for (const command of commands) {
      let processedCommand = command;
      for (const [key, value] of Object.entries(variables)) {
        processedCommand = processedCommand.replace(new RegExp(`\\$${key}`, 'g'), value);
      }
      
      const tokens = this.tokenizeCommand(processedCommand);
      
      if (tokens[0] === 'cd') {
        const targetDir = tokens[1];
        if (targetDir.startsWith('/')) {
          currentDir = targetDir;
        } else {
          currentDir = require('path').resolve(currentDir, targetDir);
        }
        results.push({ command: processedCommand, type: 'cd', result: { cwd: currentDir } });
      } else if (tokens[0] === 'lql' || tokens[0] === 'launchql') {
        // Handle LaunchQL CLI commands
        const argv = this.parseCliCommand(tokens.slice(1), currentDir);
        if (executeCommands) {
          const result = await this.runCliCommand(argv);
          results.push({ command: processedCommand, type: 'cli', result });
        } else {
          results.push({ command: processedCommand, type: 'cli', result: { argv } });
        }
      } else {
        throw new Error(`Unsupported command: ${tokens[0]}`);
      }
    }
    
    return results;
  }

  private tokenizeCommand(command: string): string[] {
    const re = /("[^"\\]*(?:\\.[^"\\]*)*"|'[^'\\]*(?:\\.[^'\\]*)*'|--[a-zA-Z0-9_-]+|-[a-zA-Z0-9_]+|[^\s]+)|(\s+)/g;
    const matches = command.match(re);
    const tokens = matches ? matches.filter(t => t.trim()) : [];
    return tokens;
  }

  private parseCliCommand(tokens: string[], cwd: string): ParsedArgs {
    const argv: ParsedArgs = {
      _: [],
      cwd
    };
    
    let i = 0;
    while (i < tokens.length) {
      const token = tokens[i];
      
      if (token.startsWith('--')) {
        const key = token.substring(2);
        
        // Handle --no-* flags by setting the base property to false
        if (key.startsWith('no-')) {
          const baseKey = key.substring(3);
          argv[baseKey] = false;
          i += 1;
        } else if (i + 1 < tokens.length && !tokens[i + 1].startsWith('--')) {
          argv[key] = tokens[i + 1];
          i += 2;
        } else {
          argv[key] = true;
          i += 1;
        }
      } else {
        argv._.push(token);
        i += 1;
      }
    }
    
    return argv;
  }

  private async runCliCommand(argv: ParsedArgs): Promise<any> {
    const prompter = new Inquirerer({
      input: process.stdin,
      output: process.stdout,
      noTty: true
    });

    const options: CLIOptions & { skipPgTeardown?: boolean } = {
      noTty: true,
      input: process.stdin,
      output: process.stdout,
      version: '1.0.0',
      skipPgTeardown: true,
      minimistOpts: {
        alias: {
          v: 'version',
          h: 'help'
        }
      }
    };

    await commands(argv, prompter, options);

    return { argv };
  }

  async cleanup(): Promise<void> {
    // Don't close pools here - let pg-cache manage them
    this.pools = [];
    this.databases = [];
    
    super.cleanup();
  }
}
