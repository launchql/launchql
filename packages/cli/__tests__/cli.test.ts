import { Inquirerer, Question } from 'inquirerer';
import { KEY_SEQUENCES, setupTests, TestEnvironment } from '../test-utils';

const beforeEachSetup = setupTests();

describe('Inquirerer', () => {
  let environment: TestEnvironment;

  beforeEach(() => {
    environment = beforeEachSetup();
  });

  it('prompts user and correctly processes delayed input', async () => {
   const { mockInput, mockOutput, writeResults, transformResults, enqueueInputResponse } = environment;

    const prompter = new Inquirerer({
      input: mockInput,
      output: mockOutput,
      noTty: false
    });

    const questions: Question[] = [{
      name: 'autocompleteField',
      type: 'autocomplete',
      options: ['first option', 'firry second option', 'firry third option']
    }];

    const argv = {};

    enqueueInputResponse({ type: 'read', value: 'fir' });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.ENTER });

    const result = await prompter.prompt(argv, questions);

    expect(argv).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(writeResults).toMatchSnapshot();
    expect(transformResults).toMatchSnapshot();
  });
});
