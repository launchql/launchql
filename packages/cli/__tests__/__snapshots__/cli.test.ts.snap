// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inquirerer prompts user and correctly processes delayed input 1`] = `
{
  "autocompleteField": "firry third option",
}
`;

exports[`Inquirerer prompts user and correctly processes delayed input 2`] = `
{
  "autocompleteField": "firry third option",
}
`;

exports[`Inquirerer prompts user and correctly processes delayed input 3`] = `
[
  "",
  "autocompleteField?
Argument<SPACE>--autocompleteField<SPACE>type<SPACE>[autocomplete]
><SPACE>Your<SPACE>input:
$<SPACE>
",
  "><SPACE>first<SPACE>option
",
  "<SPACE><SPACE>firry<SPACE>second<SPACE>option
",
  "<SPACE><SPACE>firry<SPACE>third<SPACE>option
",
  "",
  "autocompleteField?
Argument<SPACE>--autocompleteField<SPACE>type<SPACE>[autocomplete]
><SPACE>Your<SPACE>input:
$<SPACE>
",
  "<SPACE><SPACE>first<SPACE>option
",
  "><SPACE>firry<SPACE>second<SPACE>option
",
  "<SPACE><SPACE>firry<SPACE>third<SPACE>option
",
  "",
  "autocompleteField?
Argument<SPACE>--autocompleteField<SPACE>type<SPACE>[autocomplete]
><SPACE>Your<SPACE>input:
$<SPACE>
",
  "<SPACE><SPACE>first<SPACE>option
",
  "<SPACE><SPACE>firry<SPACE>second<SPACE>option
",
  "><SPACE>firry<SPACE>third<SPACE>option
",
]
`;

exports[`Inquirerer prompts user and correctly processes delayed input 4`] = `
[
  "",
  "",
  "<ENTER>",
]
`;
